toreview/BitSetUtils.java

```diff
@@ -0,0 +1,174 @@
+package com.gruppo1.distributednetworkmanager;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.BitSet;
+
+public class BitSetUtils {
+
+    private static int minLength = 64;
+    private static final String NOT_VALID_NUMBIT_EXCEPTION_MSG = "numBits isn't > 0 or a multiple of 64";
+    private static final String NOT_VALID_NUMBIT_SHA_EXCEPTION_MSG = "numBits isn't > 0 or <= 160";
+    private static final String INVALID_HEX_CHAR_MSG = "Invalid Hexadecimal Character: ";
+    private static final String INVALID_HEX_STRING_MSG = "Invalid hexadecimal String supplied.";
+    private static final String UTILS_TAG = BitSetUtils.class.toString();
+    private static final String SHA_1 = "SHA-1";
+
+    /**
+     * @param first the first binary key
+     * @param second the second binary key
+     * @return The distance of the Keys, calculated in XOR logic
+     */
+    public static BitSet distanceFrom(@NonNull BitSet first, @NonNull BitSet second) {
+        BitSet distance = (BitSet) first.clone();
+        distance.xor(second);
+        return distance;
+    }
+
+    /**
+     * @param toHash  the String (a key) of which generate hashcode
+     * @param numBits number of bits that the hash code will be constituted of, must be multiple of 64 and >0
+     * @return the bitSet containing the hash of the peer's address, bitSet's length is a multiple of 64
+     * @throws IllegalArgumentException if peer isn't valid or the numBit isn't multiple of 64
+     */
+    public static BitSet hash(@NonNull String toHash, int numBits) {
+        if (numBits > 0 && numBits % minLength == 0) {
+            int numLong = numBits / minLength;
+            long[] numbers = new long[numLong];
+            for (int i = 0; i < numLong; i++) {
+                numbers[numLong - (i + 1)] = Double.doubleToLongBits(hash(toHash) / (i + 1));

```

Comment:
Ricalcolate l'hash sulla stessa stringa ad ogni ciclo? Può essere ottimizzato portando quel calcolo fuori dal `for`.

Reply:
Non ho ben capito questo procedimento. Avete due metodi `hash`: uno che si serve di `MessageDigest` su un array di `byte`, l'altro che calcola un hash sulla stringa e popola un array. Perché non utilizzate `byte[] hashedkey = digestAlgorithm.digest(toHash.getBytes(StandardCharsets.UTF_8));` ? E poi da questo array potete generare un `BitSet` come avete fatto con `numbers`.

Reply:
Perchè con questo volevo ricreare l'algoritmo di hashing della classe String di java però su un numero arbitrario di bit, l'altro che prende in input il vettore di byte esegue un altro algoritmo.
Se vai a vedere sulle pagine della oracle hashcode di java.String è fatto proprio come il metodo hash(String) che trovi sotto.
per il ricalcolo dell'hash hai ragione, meglio spostarlo fuori.

Reply:
Si l'hash di una stringa è calcolato proprio cosi, ma non ho ancora capito perché non vuoi prendere in considerazione questo:
`byte[] hashedkey = digestAlgorithm.digest(toHash.getBytes(StandardCharsets.UTF_8))`?
Prima ottengo i byte della stringa di cui fare l'hash in codifica UTF-8. Poi ottengo il vero hash su questi bytes usando un `MessageDigest` che generara 256 bit. Ad esempio noi utilizziamo SHA-256 e poi lo tronchiamo in base ai bit che utilizziamo.

Reply:
Perchè è un'altro algoritmo. Validissimo, ma non era quello che volevo realizzare.
Se vedi nell'altro hash pubblico ho utilizzato l'sha-1.
In realtà andrebbe studiato bene quale algoritmo di hashing usare per la generazione delle chiavi, anche in base a quanti bit si vanno a tenere e in base a quale tipo di oggetti usano come base per il calcolo. Per esempio, stringhe lunghe? Stringhe corte?
Poi calcolare 256 bit di hash per poi usarne meno della metà può essere uno spreco, e ancora dipende da che tipo di oggetto parti

Reply:
Beh anche voi lo troncate mi sembra

Reply:
Sì ma da 160 bit

Reply:
Comunque dipende dall'utilizzo, non mi sento di dire che un'opzione sia meglio o peggio.